### PR TITLE
add simple fsm example

### PR DESCRIPTION
**Pull request by alexpantyukhin**
Just small example for demonstrating fsm.

### CODE REVIEW

**Comment by cpx86** [comment]
Cool stuff! I wonder though if maybe we should separate this example from the others. The examples we have so far are mainly intended as an executable documentation of protoactor features, whereas this is more of an example of how an FSM pattern can be implemented using actors.

E.g. perhaps the `examples` directory could be split into something like:
`examples/features`
`examples/benchmarks`
`examples/patterns`

@AsynkronIT/developers thoughts?

**Comment by alexpantyukhin** [comment]
@cpx86 I agree with your suggestion about splitting examples types.

**Comment by alexpantyukhin** [comment]
The example is based on the akka.net implementation and it is simplfied a little. I think I can simplify it more. But now it is working prototype :)

**Comment by cpx86** [comment]
Would be nice to split the file as well - it was a bit hard to read the diff tbh 😄

**Comment by alexpantyukhin** [comment]
done :) cleaned up a little.

**Comment by alexpantyukhin** [comment]
@cpx86 I thought a little and got that benchmarks actually are not examples.
What if
examples/features
   Chat
   ClusterHelloWorld
   DependencyInjection
   Futures
   HelloWorld
   Labb
   LifecycleEvents
   Middleware
   Persistence
   ReceiveTimeout
   RemoteActivate
   Router
   Schedulers/SimpleScheduler
   Supervision

benchmarks
   InprocessBenchmark
   MailboxBenchmark
   RemoteBenchmark
   SpawnBenchmark

examples/patterns
   FSM

**Comment by rogeralsing** [comment]
While I do think FSM DSL's can be useful, I would like to bring this one up on the design table and see if we can make this more C# 7 idiomatic and how it fits with the ideas around Proto.Actor.

Thoughts @AsynkronIT/developers ?

**Comment by adamhathcock** [comment]
I would rather see how something like Stateless would be used instead of making something new.  I've liked using it.  Maybe a tighter integration could be done.

https://github.com/dotnet-state-machine/stateless/

**Comment by rogeralsing** [comment]
Does Stateless allow for event triggers, e.g. timeouts or how can we integrate something like this in the actor world?

But I agree, that is exactly the Proto.Actor way to go about things. re-use proven building blocks

**Comment by adamhathcock** [comment]
You can definitely trigger events on state transitions. Not sure what you mean about timeouts.

I might get some time tomorrow to implement a FSM example.

**Comment by rogeralsing** [comment]
In Akka, you can set timeouts for states, so that if you have been in state X for Y amount of time, then you should revert to some other state.

**Comment by adamhathcock** [comment]
Actually, when attempting to use Stateless, I'm basically having to recreate what's going on in this PR anyway.

Stateless's main usefulness, it seems, is to just have a graph of validate states so the current state is tracked and allows a transition to happen or not.

Most of the code in this PR is doing actor specific things.  Basically, it wasn't adding much as the state handlers could not be attached and then you add the burden of the user having to declare what state transitions are allowed (well this could be optional).

So, in conclusion.  Users might want to layer Stateless on top of this FSM for validation purposes but it doesn't add anything that I can see.

**Comment by alexpantyukhin** [comment]
Need to close this PR?
